qemu/tests/block_hotplug_negative.py

```diff
@@ -0,0 +1,108 @@
+        """
+        error_context.context("unplug the drive", logging.info)
+        drive.unplug(vm.monitor)
+        ver_out = drive.verify_hotplug("", vm.monitor)
+        time.sleep(5)
+        error_context.context("Hotplug the drive", logging.info)
+        drive.hotplug(vm.monitor)
+
+    img_list = params.get("images").split()
+    img_format_type = params.get("img_format_type", "qcow2")
+    pci_type = params.get("pci_type", "virtio-blk-pci")
+    blk_num = int(params.get("blk_num", 1))
+    add_block_device = True
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    error_context.context("Hotplug block device", logging.info)
+    for num in xrange(blk_num):
+        device = qdevices.QDevice(pci_type)
+        drive = qdevices.QRHDrive("block%d" % num)
+        drive.set_param("file", find_image(img_list[num + 1]))
+        drive.set_param("format", img_format_type)
+        drive_id = drive.get_param("id")
+        drive.hotplug(vm.monitor)
+        #add controller if needed
+        if params.get("need_controller", "no") == "yes":
+            controller_model = params.get("controller_model")
+            if params.get("bus_extra_params_%s" % img_list[num + 1]):
+                bus_extra_param = params["bus_extra_params_%s" % img_list[num + 1]]
+                match = re.search("(\w+)=(\w+)", bus_extra_param)
+                if match:
+                    qdevice_params = {match.group(1): match.group(2)}
+                    controller = qdevices.QDevice(controller_model, params=qdevice_params)
+            else:
+                controller = qdevices.QDevice(controller_model)
+            try:
+                controller.hotplug(vm.monitor)
+            except Exception, e:
+                if "QMP command 'device_add' failed" in str(e):
+                    error_context.context("Failed to add controller with invalid params", logging.info)
+                    drive_unplug_plug(drive, vm)
+                    add_block_device = False
+
+        if add_block_device:
+            device.set_param("drive", drive_id)
+            device.set_param("id", "block%d" % num)
+            if params.get("blk_extra_params_%s" % img_list[num + 1]):
+                blk_extra_param = params["blk_extra_params_%s" % img_list[num + 1]]
+                match = re.search("(\w+)=(\w+)", blk_extra_param)
+                if match:
+                    device.set_param(match.group(1), match.group(2))
+            try:
+                device.hotplug(vm.monitor)
+            except Exception, e:
+                if "QMP command 'device_add' failed" in str(e):
+                    error_context.context("Failed to add block with invalid params",
+                                          logging.info)
+                    drive_unplug_plug(drive, vm)
+
+    error_context.context("Check vm is alive after drive unplug/hotplug test", logging.info)
+    session = vm.wait_for_login()
+    if not session.is_responsive():
```

Comment:
please use:
try:
     ****
except:
    ****
finally:
    ****

Reply:
No need to use this pattern, for session.is_responsive() returns  "True or False".

Reply:
actually, my point is session.close() in 'finally'.